packages/markdowndb/src/lib/markdowndb.ts

```diff
@@ -1,281 +1,253 @@
+      const urlPath = sluggifyFilePath(pathRelativeToFolder);
+      fileToInsert.url_path = urlPath;
+
+      // extension
+      const [, extension] = filePath.match(/.(\w+)$/) || [];
+      fileToInsert.extension = extension;
+
+      if (!File.supportedExtensions.includes(extension)) {
+        filesToInsert.push(fileToInsert);
+        continue;
+      }
 
-  //  If it's not a md/mdx file, _url_path is just the relative path
-  const pathRelativeToFolder = path.slice(folderPath.length + 1);
-  let _url_path = pathRelativeToFolder;
+      // metadata, tags, links
+      const source: string = fs.readFileSync(filePath, {
+        encoding: "utf8",
+        flag: "r",
+      });
 
-  if (["md", "mdx"].includes(filetype)) {
-    const source = fs.readFileSync(path, { encoding: "utf8", flag: "r" });
-    const { data } = matter(source);
-    metadata = data || null;
-    type = data.type || null;
+      const { metadata, links } = parseFile(source);
+      fileToInsert.filetype = metadata?.type || null;
+      fileToInsert.metadata = metadata || null;
 
-    const segments = pathRelativeToFolder.split("/");
-    const filename = segments.at(-1).split(".")[0];
+      // TODO is there a better way to do this?
+      filesLinksMap[id] = {
+        url: urlPath,
+        links,
+      };
 
-    const pathToFileFolder = segments.slice(0, -1).join("/");
+      const tags = metadata?.tags || [];
+      tags.forEach((tag: string) => {
+        if (!tagsToInsert.some((t) => t.name === tag)) {
+          tagsToInsert.push({ name: tag });
+        }
+        fileTagsToInsert.push({ file: id, tag });
+      });
 
-    if (filename != "index") {
-      if (pathToFileFolder) {
-        _url_path = `${pathToFileFolder}/${filename}`;
-      } else {
-        //  The file is in the root folder
-        _url_path = filename;
-      }
-    } else {
-      _url_path = pathToFileFolder;
+      filesToInsert.push(fileToInsert);
     }
-  }
 
-  return {
-    _id: id,
-    _path: path,
-    _url_path, //  Should exist only for MD/MDX files
-    filetype,
-    metadata,
-    type,
-  };
-};
+    Object.entries(filesLinksMap).forEach(([fileId, { url, links }]) => {
+      links.forEach(({ linkSrc, linkType }) => {
+        const destPath = resolveLinkToUrlPath(linkSrc, url);
+        const destFile = filesToInsert.find(
+          (file) => file.url_path === destPath
+        );
+        const linkToInsert = {
+          // _id: id,
+          from: fileId,
+          to: destFile?._id,
+          link_type: linkType,
+        };
+        linksToInsert.push(linkToInsert);
+      });
+    });
 
-class MarkdownDB {
-  db: Knex;
+    await File.batchInsert(this.db, filesToInsert);
 
-  constructor(db: Knex) {
-    this.db = db;
+    // TODO  what happens if some of the files were not inserted?
+    // I guess inserting tags or links with such files used as foreign keys will fail too,
+    // but need to check
```

Comment:
I'm not 100% sure this is how it works, you can provide a `.then(function(ids) { /*...*/ }).catch(function(error) { /*...*/ });` to batchInsert, so maybe the error will contain the ones that failed? I think it would be better to experiment a bit